File: binarySearch/searchInRotatedSortedArray.py
class Solution(object):
    def search(self, nums, target):
        #find the pivot point k by finding index of miniumum value
        numsLen = len(nums)
        l, r = 0, numsLen - 1
        while nums[l] > nums[r]:
            m = l + (r - l) / 2
            if nums[m] > nums[r]:
                l = m + 1
            else:
                r = m
        #you can modify the indexes with the pivot value k to do a classical binary search
        k = (numsLen - l)  % numsLen
        mapped = lambda x: (x + k) % numsLen
        unMapped = lambda x: (x - k + numsLen) if (x - k) < 0 else (x - k)
        r = unMapped(mapped(l) + numsLen - 1)
        while mapped(l) <= mapped(r):
            m = mapped(l) + (mapped(r) - mapped(l)) / 2
            if target < nums[unMapped(m)]:
                if m - 1 < 0:
                    break
                r = unMapped(m - 1)
            elif target > nums[unMapped(m)]:
                if m + 1 > numsLen - 1:
                    break
                l = unMapped(m + 1)
            else:
                return unMapped(m)
        return -1

File: binarySearch/searchInRotatedSortedArray.cpp
class Solution {
public:
    int rot_num(vector<int> nums){
        int l = 0;
        int r = nums.size() - 1;
        while (l < r){
            int m = l + (r - l) / 2;
            if (nums[m] > nums[r])
                l = m + 1;
            else
                r = m;
        }
        return l;
    }

    //create mapping function using the number of rotations
    int map(int index, int rots, int arr_size){
        return (index + rots) % arr_size;
    }

    int search(vector<int>& nums, int target) {
        //find pivot point or number of rotations (find min in rotated sorted array)
        int rots = rot_num(nums);
        int arr_size = nums.size();

        //now, you can do a nominal regular binary search using a mapped m index nums vals to adjust l and r :) 
        //unmapped indexes 
        int l = 0;
        int r = nums.size() - 1;
        while (l <= r){
            int m = l + (r - l) / 2;
            int m_m = map(m, rots, arr_size);
            if(target < nums[m_m])
                r = m - 1;
            else if(target > nums[m_m])
                l = m + 1;
            else
                return m_m;
        }
        return -1;
    }
};
